File: Tests/RedlineCoreTests/FleetTests.swift
// Exercises the Claude Code session registry reader against synthetic records in a temp dir.
import XCTest
@testable import RedlineCore

final class ClaudeFleetStoreTests: XCTestCase {
    private var dir: URL!
    /// A fixed start time every fixture claims, so a record and the probe agree by default
    private let started = Date(timeIntervalSince1970: 1_787_067_508)

    override func setUpWithError() throws {
        dir = URL(fileURLWithPath: NSTemporaryDirectory())
            .appendingPathComponent("redline-fleet-" + UUID().uuidString)
        try FileManager.default.createDirectory(at: dir, withIntermediateDirectories: true)
    }

    override func tearDownWithError() throws {
        try? FileManager.default.removeItem(at: dir)
    }

    private func ctime(_ d: Date) -> String {
        let f = DateFormatter()
        f.locale = Locale(identifier: "en_US_POSIX")
        f.dateFormat = "EEE MMM d HH:mm:ss yyyy"
        return f.string(from: d)
    }

    private func write(_ obj: [String: Any], pid: Int) throws {
        let data = try JSONSerialization.data(withJSONObject: obj)
        try data.write(to: dir.appendingPathComponent("\(pid).json"))
    }

    private func record(pid: Int, status: String, statusAt: Date,
                        extra: [String: Any] = [:]) -> [String: Any] {
        var obj: [String: Any] = [
            "pid": pid,
            "sessionId": "s-\(pid)",
            "cwd": "/Users/x/work/proj-\(pid)",
            "startedAt": Int(started.timeIntervalSince1970 * 1000),
            "procStart": ctime(started),
            "version": "2.1.234",
            "kind": "interactive",
            "entrypoint": "cli",
            "name": "proj-\(pid)",
            "status": status,
            "updatedAt": Int(statusAt.timeIntervalSince1970 * 1000),
            "statusUpdatedAt": Int(statusAt.timeIntervalSince1970 * 1000),
            "bridgeSessionId": "session_\(pid)",
        ]
        for (k, v) in extra { obj[k] = v }
        return obj
    }

    /// Every PID alive, all claiming the same start the fixtures write
    private var allAlive: ProcessProbe {
        ProcessProbe { [started] _ in started }
    }

    private func store(_ probe: ProcessProbe? = nil) -> ClaudeFleetStore {
        ClaudeFleetStore(root: dir, probe: probe ?? allAlive)
    }

    func testReadsAWaitingSessionWithWaitingFor() throws {
        let now = Date(timeIntervalSince1970: 1_787_070_000)
        try write(record(pid: 100, status: "waiting",
                         statusAt: now.addingTimeInterval(-840),
                         extra: ["waitingFor": "input needed"]), pid: 100)
        let snap = store().scan(now: now)
        XCTAssertEqual(snap.sessions.count, 1)
        let s = snap.sessions[0]
        XCTAssertEqual(s.state, .waiting)
        XCTAssertEqual(s.waitingFor, "input needed")
        XCTAssertEqual(s.folder, "proj-100")
        XCTAssertEqual(s.label, "proj-100")
        XCTAssertEqual(s.timeInStatus(now: now).map { Int($0) }, 840)
        XCTAssertEqual(s.claudeURL?.absoluteString,
                       "https://claude.ai/code/session_100")
        XCTAssertEqual(snap.waiting.count, 1)
    }

    func testUnknownEntrypointAndUnknownFieldsSurvive() throws {
        let now = Date(timeIntervalSince1970: 1_787_070_000)
        try write(record(pid: 101, status: "busy", statusAt: now,
                         extra: ["entrypoint": "desktop",
                                 "kind": "headless",
                                 "someFutureField": ["nested": true],
                                 "peerProtocol": 7]), pid: 101)
        let snap = store().scan(now: now)
        XCTAssertEqual(snap.sessions.count, 1)
        XCTAssertEqual(snap.sessions[0].entrypoint, "desktop")
        XCTAssertEqual(snap.sessions[0].kind, "headless")
        XCTAssertEqual(snap.sessions[0].state, .busy)
    }

    func testAnUnknownStatusStillShowsTheSession() throws {
        let now = Date(timeIntervalSince1970: 1_787_070_000)
        try write(record(pid: 102, status: "compacting", statusAt: now), pid: 102)
        let snap = store().scan(now: now)
        XCTAssertEqual(snap.sessions.count, 1)
        XCTAssertEqual(snap.sessions[0].state, .unknown)
        XCTAssertEqual(snap.sessions[0].status, "compacting")
    }

    func testMalformedRecordDoesNotTakeOutItsNeighbours() throws {
        let now = Date(timeIntervalSince1970: 1_787_070_000)
        try write(record(pid: 200, status: "idle", statusAt: now), pid: 200)
        try "{not json at all".write(to: dir.appendingPathComponent("201.json"),
                                    atomically: true, encoding: .utf8)
        // Valid JSON, but missing the two fields a row cannot be drawn without
        try write(["status": "busy"], pid: 202)
        let snap = store().scan(now: now)
        XCTAssertEqual(snap.sessions.map { $0.pid }, [200])
    }

    func testDeadPidIsTreatedAsAbsentAndTheFileIsLeftAlone() throws {
        let now = Date(timeIntervalSince1970: 1_787_070_000)
        try write(record(pid: 300, status: "waiting", statusAt: now), pid: 300)
        let snap = ClaudeFleetStore(root: dir, probe: ProcessProbe { _ in nil }).scan(now: now)
        XCTAssertTrue(snap.isEmpty)
        XCTAssertTrue(FileManager.default.fileExists(
            atPath: dir.appendingPathComponent("300.json").path))
    }

    /// Claude Code writes procStart in UTC while the field names no zone, so reading it as
    /// local rejected every live session on a machine west of Greenwich. Both readings count.
    func testProcStartWrittenInUTCIsAccepted() throws {
        let now = Date(timeIntervalSince1970: 1_787_070_000)
        var obj = record(pid: 302, status: "busy", statusAt: now)
        let utc = DateFormatter()
        utc.locale = Locale(identifier: "en_US_POSIX")
        utc.dateFormat = "EEE MMM d HH:mm:ss yyyy"
        utc.timeZone = TimeZone(secondsFromGMT: 0)
        obj["procStart"] = utc.string(from: started)
        try write(obj, pid: 302)
        XCTAssertEqual(store().scan(now: now).sessions.map { $0.pid }, [302])
    }

    /// PIDs are reused, so a record whose claimed start does not match the live process is a
    /// leftover pointing at somebody else's process
    func testRecycledPidIsRejected() throws {
        let now = Date(timeIntervalSince1970: 1_787_070_000)
        try write(record(pid: 301, status: "busy", statusAt: now), pid: 301)
        let other = started.addingTimeInterval(3600)
        let snap = ClaudeFleetStore(root: dir,
                                    probe: ProcessProbe { _ in other }).scan(now: now)
        XCTAssertTrue(snap.isEmpty)
    }

    func testSortsWaitingFirstThenLongestInStatus() throws {
        let now = Date(timeIntervalSince1970: 1_787_070_000)
        try write(record(pid: 1, status: "idle", statusAt: now.addingTimeInterval(-9000)),
                  pid: 1)
        try write(record(pid: 2, status: "busy", statusAt: now.addingTimeInterval(-60)), pid: 2)
        try write(record(pid: 3, status: "waiting", statusAt: now.addingTimeInterval(-300)),
                  pid: 3)
        try write(record(pid: 4, status: "waiting", statusAt: now.addingTimeInterval(-3600)),
                  pid: 4)
        try write(record(pid: 5, status: "busy", statusAt: now.addingTimeInterval(-600)), pid: 5)
        let snap = store().scan(now: now)
        XCTAssertEqual(snap.sessions.map { $0.pid }, [4, 3, 5, 2, 1])
    }

    func testKeyFilesAreNeverRead() throws {
        let now = Date(timeIntervalSince1970: 1_787_070_000)
        try write(record(pid: 400, status: "idle", statusAt: now), pid: 400)
        try "supersecret".write(
            to: dir.appendingPathComponent("400.abc123.key"),
            atomically: true, encoding: .utf8)
        let snap = store().scan(now: now)
        XCTAssertEqual(snap.sessions.count, 1)
    }

    func testMissingDirectoryIsNotAnError() {
        let gone = dir.appendingPathComponent("nope")
        XCTAssertTrue(ClaudeFleetStore(root: gone).scan().isEmpty)
    }

    func testNameFallsBackToTheFolder() throws {
        let now = Date(timeIntervalSince1970: 1_787_070_000)
        var obj = record(pid: 500, status: "busy", statusAt: now)
        obj.removeValue(forKey: "name")
        obj.removeValue(forKey: "statusUpdatedAt")
        obj.removeValue(forKey: "updatedAt")
        try write(obj, pid: 500)
        let snap = store().scan(now: now)
        XCTAssertEqual(snap.sessions[0].label, "proj-500")
        XCTAssertNil(snap.sessions[0].timeInStatus(now: now))
    }

    /// The real probe, against this very process: a start time that exists and is not absurd
    func testLiveProbeAnswersForThisProcess() {
        let start = ProcessProbe.live.startTime(getpid())
        XCTAssertNotNil(start)
        if let start {
            XCTAssertLessThan(abs(start.timeIntervalSinceNow), 86400)
        }
        XCTAssertNil(ProcessProbe.live.startTime(Int32.max))
    }
}

File: windows/tests/Redline.Core.Tests/FleetTests.cs
// Exercises the Claude Code session registry reader against synthetic records in a temp dir.
using System.Globalization;
using System.Text.Json.Nodes;
using Redline.Core;

namespace Redline.Core.Tests;

public sealed class ClaudeFleetStoreTests : IDisposable
{
    private readonly string _dir;
    /// A fixed start time every fixture claims, so a record and the probe agree by default
    private static readonly DateTimeOffset Started = DateTimeOffset.FromUnixTimeSeconds(1_787_067_508);
    private static readonly DateTimeOffset Now = DateTimeOffset.FromUnixTimeSeconds(1_787_070_000);

    public ClaudeFleetStoreTests()
    {
        _dir = Path.Combine(Path.GetTempPath(), "redline-fleet-" + Guid.NewGuid());
        Directory.CreateDirectory(_dir);
    }

    public void Dispose()
    {
        try { Directory.Delete(_dir, recursive: true); } catch { }
    }

    private static string Ctime(DateTimeOffset d, bool utc = false)
    {
        var t = utc ? d.ToUniversalTime() : TimeZoneInfo.ConvertTime(d, TimeZoneInfo.Local);
        return t.ToString("ddd MMM d HH:mm:ss yyyy", CultureInfo.InvariantCulture);
    }

    private void Write(JsonObject obj, int pid) =>
        File.WriteAllText(Path.Combine(_dir, $"{pid}.json"), obj.ToJsonString());

    private static JsonObject Record(int pid, string status, DateTimeOffset statusAt, JsonObject? extra = null)
    {
        var obj = new JsonObject
        {
            ["pid"] = pid,
            ["sessionId"] = $"s-{pid}",
            ["cwd"] = $"/Users/x/work/proj-{pid}",
            ["startedAt"] = Started.ToUnixTimeMilliseconds(),
            ["procStart"] = Ctime(Started),
            ["version"] = "2.1.234",
            ["kind"] = "interactive",
            ["entrypoint"] = "cli",
            ["name"] = $"proj-{pid}",
            ["status"] = status,
            ["updatedAt"] = statusAt.ToUnixTimeMilliseconds(),
            ["statusUpdatedAt"] = statusAt.ToUnixTimeMilliseconds(),
            ["bridgeSessionId"] = $"session_{pid}",
        };
        if (extra is not null)
            foreach (var (k, v) in extra) obj[k] = v?.DeepClone();
        return obj;
    }

    /// Every PID alive, all claiming the same start the fixtures write
    private static ProcessProbe AllAlive => new(_ => Started);

    private ClaudeFleetStore Store(ProcessProbe? probe = null) => new(_dir, probe ?? AllAlive);

    [Fact]
    public void ReadsAWaitingSessionWithWaitingFor()
    {
        Write(Record(100, "waiting", Now.AddSeconds(-840), new JsonObject { ["waitingFor"] = "input needed" }), 100);
        var snap = Store().Scan(Now);
        Assert.Single(snap.Sessions);
        var s = snap.Sessions[0];
        Assert.Equal(FleetState.Waiting, s.State);
        Assert.Equal("input needed", s.WaitingFor);
        Assert.Equal("proj-100", s.Folder);
        Assert.Equal("proj-100", s.Label);
        Assert.Equal(840, s.TimeInStatus(Now) is { } t ? (int)t : (int?)null);
        Assert.Equal("https://claude.ai/code/session_100", s.ClaudeUrl?.AbsoluteUri);
        Assert.Single(snap.Waiting);
    }

    [Fact]
    public void UnknownEntrypointAndUnknownFieldsSurvive()
    {
        Write(Record(101, "busy", Now, new JsonObject
        {
            ["entrypoint"] = "desktop",
            ["kind"] = "headless",
            ["someFutureField"] = new JsonObject { ["nested"] = true },
            ["peerProtocol"] = 7,
        }), 101);
        var snap = Store().Scan(Now);
        Assert.Single(snap.Sessions);
        Assert.Equal("desktop", snap.Sessions[0].Entrypoint);
        Assert.Equal("headless", snap.Sessions[0].Kind);
        Assert.Equal(FleetState.Busy, snap.Sessions[0].State);
    }

    [Fact]
    public void AnUnknownStatusStillShowsTheSession()
    {
        Write(Record(102, "compacting", Now), 102);
        var snap = Store().Scan(Now);
        Assert.Single(snap.Sessions);
        Assert.Equal(FleetState.Unknown, snap.Sessions[0].State);
        Assert.Equal("compacting", snap.Sessions[0].Status);
    }

    [Fact]
    public void MalformedRecordDoesNotTakeOutItsNeighbours()
    {
        Write(Record(200, "idle", Now), 200);
        File.WriteAllText(Path.Combine(_dir, "201.json"), "{not json at all");
        // Valid JSON, but missing the two fields a row cannot be drawn without
        Write(new JsonObject { ["status"] = "busy" }, 202);
        var snap = Store().Scan(Now);
        Assert.Equal(new[] { 200 }, snap.Sessions.Select(s => s.Pid));
    }

    [Fact]
    public void DeadPidIsTreatedAsAbsentAndTheFileIsLeftAlone()
    {
        Write(Record(300, "waiting", Now), 300);
        var snap = new ClaudeFleetStore(_dir, new ProcessProbe(_ => null)).Scan(Now);
        Assert.True(snap.IsEmpty);
        Assert.True(File.Exists(Path.Combine(_dir, "300.json")));
    }

    /// Claude Code writes procStart in UTC while naming no zone, so reading it as local rejected
    /// every live session west of Greenwich. Both readings count.
    [Fact]
    public void ProcStartWrittenInUTCIsAccepted()
    {
        var obj = Record(302, "busy", Now);
        obj["procStart"] = Ctime(Started, utc: true);
        Write(obj, 302);
        Assert.Equal(new[] { 302 }, Store().Scan(Now).Sessions.Select(s => s.Pid));
    }

    /// PIDs are reused, so a record whose claimed start does not match the live process is a
    /// leftover pointing at somebody else's process
    [Fact]
    public void RecycledPidIsRejected()
    {
        Write(Record(301, "busy", Now), 301);
        var other = Started.AddSeconds(3600);
        var snap = new ClaudeFleetStore(_dir, new ProcessProbe(_ => other)).Scan(Now);
        Assert.True(snap.IsEmpty);
    }

    [Fact]
    public void SortsWaitingFirstThenLongestInStatus()
    {
        Write(Record(1, "idle", Now.AddSeconds(-9000)), 1);
        Write(Record(2, "busy", Now.AddSeconds(-60)), 2);
        Write(Record(3, "waiting", Now.AddSeconds(-300)), 3);
        Write(Record(4, "waiting", Now.AddSeconds(-3600)), 4);
        Write(Record(5, "busy", Now.AddSeconds(-600)), 5);
        var snap = Store().Scan(Now);
        Assert.Equal(new[] { 4, 3, 5, 2, 1 }, snap.Sessions.Select(s => s.Pid));
    }

    [Fact]
    public void KeyFilesAreNeverRead()
    {
        Write(Record(400, "idle", Now), 400);
        File.WriteAllText(Path.Combine(_dir, "400.abc123.key"), "supersecret");
        var snap = Store().Scan(Now);
        Assert.Single(snap.Sessions);
    }

    [Fact]
    public void MissingDirectoryIsNotAnError()
    {
        Assert.True(new ClaudeFleetStore(Path.Combine(_dir, "nope")).Scan().IsEmpty);
    }

    [Fact]
    public void NameFallsBackToTheFolder()
    {
        var obj = Record(500, "busy", Now);
        obj.Remove("name");
        obj.Remove("statusUpdatedAt");
        obj.Remove("updatedAt");
        Write(obj, 500);
        var snap = Store().Scan(Now);
        Assert.Equal("proj-500", snap.Sessions[0].Label);
        Assert.Null(snap.Sessions[0].TimeInStatus(Now));
    }

    /// Windows has no controlling terminal device, so the probe answers "none" for any PID,
    /// which is the macOS answer for a process without a tty.
    [Fact]
    public void TTYPathIsADeviceOrNothing()
    {
        var tty = ProcessProbe.TtyPath(Environment.ProcessId);
        if (tty is not null)
        {
            Assert.StartsWith("/dev/", tty);
            Assert.True(tty.Length > 5);
        }
        Assert.Null(ProcessProbe.TtyPath(int.MaxValue));
    }

    /// The real probe, against this very process: a start time that exists and is not absurd
    [Fact]
    public void LiveProbeAnswersForThisProcess()
    {
        var start = ProcessProbe.Live.StartTime(Environment.ProcessId);
        Assert.NotNull(start);
        if (start is { } s) Assert.True(Math.Abs((DateTimeOffset.UtcNow - s).TotalSeconds) < 86400);
        Assert.Null(ProcessProbe.Live.StartTime(int.MaxValue));
    }
}
